### ISSUE TITLE
Introduce separate annotations for Path/Query/Databinding params

### ISSUE THREAD

**Issue by chamil321** [ballerina-platform/ballerina-lang#21303]
**Description:**
Add @http:bodyparam annotation in the resource signature when binding payload instead of specifying the body field in the resource config.
Add @http:queryparam annotation in the resource signature instead of having req.getQueryParamValue(string key) function.
Evaluate the possibility of having @http:pathparam annotation and making it work without path field in the resource config.

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by chamil321** [comment]
With this change, we will be introducing query params in the resource function signature. ATM we have basically 3 attached functions of http: Request object to retrieve query params.

```ballerina
    # Gets the query parameters of the request, as a map of string array.
    #
    # + return - String array map of query params
    public function getQueryParams() returns map<string[]> = external;

    # Gets the query param value associated with the given key.
    #
    # + key - Represents query param key
    # + return - Query param value associated with the given key as a string. If multiple param values are
    #            present, then the first value is returned. Nil is returned if no key is found.
    public function getQueryParamValue(@untainted string key) returns @tainted string?;

    # Gets all the query param values associated with the given key.
    #
    # + key - Represents query param key
    # + return - All the query param values associated with the given key as a `string[]`. Nil is returned if no key
    #            is found.
    public function getQueryParamValues(@untainted string key) returns @tainted string[]?;
```

The approach was to keep the first API which returns a map of string array as it is to get all param if the user requires. So that they have the control to manipulate them upon their requirement. 
Then the second and the third APIs will be removed and moved to resource signature, the key will be the parameter name and the type of can be either string[] or string. This is not a union type. in fact separate types. We validate it at the compile time

```ballerina
string[]? bar = req.getQueryParamValues("bar"); ==>> @http:QueryParam string[] bar
string? foo = req.getQueryParamValue("foo");     ==>> @http:QueryParam string foo
```
So far the task is completed to achieve above, But following is a suggestion

Improve the type of query params to be **string[]?** or **string?**. So if the type is string[]? we return string array and if the query is not present we return nil.
   Pros : Resource executes even if the query param is not available(Bcz ATM if the query is not present in the URL, an error response is sent back to the client saying Bad Request)
   Cons : user should do a type check to differentiate the type of parameter

Please share your thoughts!!
@shafreenAnfar @wggihan @daneshk

**Comment by chamil321** [comment]
The conclusion from the code review:

- Keep the existing query param related functions as it is.
- Use **string** and **string[]** as query param types in the signature(No need of union types)

**Comment by chamil321** [comment]
Fixed by #22058, hence closing PR
